src/aae_controller.erl

```diff
@@ -811,7 +842,64 @@ handle_call({fetch_clocks, IndexNs, SegmentIDs, ReturnFun, PreflFun},
         _ ->
             {reply, ok, S0}
     end;
+handle_call({fetch_clocks,
+                IndexNs, 
+                RLimiter, SegmentIDs, LMDLimiter, 
+                ReturnFun, PreflFun},
+                    From, State) ->
+    ImmediateReply = State#state.parallel_keystore,
+    case ImmediateReply of
+        true ->
+            %% When this is a parallel store, there is no need benefit on
+            %% waiting before replying - as no race conditions will be
+            %% avoided.
+            gen_server:reply(From, ok);
+        _ ->
+            %% In native mode, don't want a new PUT to be received before the
+            %% snapshot has been taken, so block until the snapshot has been
+            %% taken
+            ok
+    end,
+    
+    FoldObjFun = 
+        fun(B, K, V, Acc) ->
+            {clock, VC} = lists:keyfind(clock, 1, V),
+            [{B, K, VC}|Acc]
+        end,
+    WrappedFoldObjFun = preflist_wrapper_fun(FoldObjFun, IndexNs),
+    SizeFun =
+        fun(KeyClockList) ->
+            length(KeyClockList)
+        end,
 
+    ok = maybe_flush_puts(State#state.key_store, 
+                            State#state.objectspecs_queue,
+                            State#state.parallel_keystore,
+                            true),
+    {async, Folder} = 
+        aae_keystore:store_fold(State#state.key_store, 
+                                RLimiter,
+                                {segments, SegmentIDs, ?TREE_SIZE},
+                                LMDLimiter,
+                                false, 
+                                WrappedFoldObjFun, 
+                                [],
+                                [{preflist, PreflFun}, 
+                                    {clock, null},
+                                    {aae_segment, null}]),
+    Queue = State#state.runner_queue ++ [{work, Folder, ReturnFun, SizeFun}],
+    Backlog = length(Queue) >= ?MAX_RUNNER_QUEUEDEPTH,
+    %% If there is a backlog, the queue will remain in backlog state until it
+    %% is empty.  Whilst in backlog state, query requests will result in dummy
+    %% work being added to the queue not real work (which may cause the queue
+    %% to continue to grow)
+    S0 = State#state{queue_backlog = Backlog, runner_queue = Queue},
+    case ImmediateReply of
+        true ->
```

Comment:
This is tricky. In case of Immediate reply, you make sure the calling process can continue, and clearly you don't want to reply once more, since the client is not waiting on the reply any more.
But whether or not this can lead to eager clients overloading this server is unclear to me.

Reply:
Having thought about this more, I think the ImmediateReply is a premature optimisation, and an unnecessary branch of logic.  It is implemented as there is no need to make the calling service wait for the flush_puts and snapshot to occur when in parallel mode - however, this is only called by exchanges, there is no reason to assume  they would benefits from a 1ms quicker response.  I think I will cut the differentiation on reply for parallel/native stores.

Reply:
Actually, I recall, the reason for ImmediateReply is that these requests are routed via the vnode, and so are blocking the vnode - so a faster reply frees up the vnode for processing sooner.  However, there will only be a delay if there is a backlog in the AAE system, and that must always eventually slow the vnode down anyway.

Although partly justifiable, I think overall this difference in behaviour between native/parallel is unhelpful additional cognitive load for the future developer.

Reply:
The ImmediateReply was added in specifically because of an issue - https://github.com/martinsumner/kv_index_tictactree/pull/84/files.  In native mode, there are no PUTs to send to the keystore, as there is no parallel keystore.  So in native mode there is no risk of the call to fetch_clocks hitting a timeout, and hence crashing the vnode.  However, the risk of a build of PUTs exists in parallel mode.  

So the distinction in behaviour is important.  

I will keep ImmediateReply, and update the comments to reflect this ... save me from going through this loop again.